src/pyhf/pdf.py

```diff
@@ -701,6 +707,12 @@ def expected_auxdata(self, pars):
         """
         tensorlib, _ = get_backend()
         pars = tensorlib.astensor(pars)
+        # Verify parameter shapes
+        if pars.shape[-1] != self.config.npars:
+            raise exceptions.InvalidPdfParameters(
+                f'eval failed as pars has len {pars.shape[-1]} but {self.config.npars} was expected'
+            )
+
         return self.make_pdf(pars)[1].expected_data()
 
     def _modifications(self, pars):
```

Comment:
given that we do input validation here it suggests that this has become a publicly consumable API. maybe we should add a `model.modifications` and do the input checks there and call `_modifications` if inputs are ok

Reply:
`_modifications` isn't currently a public API. This certainly could become one, although I might argue that we remove it from `Model` and keep it on `MainModel` unless there's a reason to pass-through it. e.g. `pdf.main_model.modifications` is just as clear to me. Unless the suggestion here is to remove the checks from `main_model` and `constraint_model` and keep all checks on `model` which is also possible, but feels like a mess.

Reply:
`main_model` is needed for `return_by_sample`, so it feels similarly "public" as `Model`.

Reply:
Sure, but `main_model.expected_actualdata(..., return_by_sample)` is fine -- since that's public. However `main_model._modifications` isn't necessarily public. Although at the moment only used by `main_model.expected_data` -- so I'm fine either way.

Reply:
> Unless the suggestion here is to remove the checks from `main_model` and `constraint_model`

My comment was meant as an example for why that may not be desirable, sorry I should have made that clear. Promoting `_modifications` to public in the longer term is a nice idea, it looks very useful for model debugging.

Reply:
@lukasheinrich @kratsg If we can revisit the performance impact here soon it would be nice to have this get into `v0.7.0`.